enterprise/server/remote_crypter/remote_crypter_test.go

```diff
@@ -0,0 +1,301 @@
+}
+
+type fakeKey struct {
+	id      keyID
+	seq     int
+	version int64
+	key     []byte
+}
+
+func newFakeEncryptionService(authenticator interfaces.Authenticator) *fakeEncryptionService {
+	return &fakeEncryptionService{
+		authenticator: authenticator,
+		keys:          make(map[groupID]map[keyID][]*fakeKey),
+	}
+}
+
+func (f *fakeEncryptionService) add(t *testing.T, group string, newKey *fakeKey) {
+	gid := groupID(group)
+	for id, keys := range f.keys[gid] {
+		for _, key := range keys {
+			if id == newKey.id && key.version == newKey.version {
+				t.FailNow()
+			}
+		}
+	}
+
+	newKey.seq = f.seq
+	f.seq++
+	if f.keys[gid] == nil {
+		f.keys[gid] = map[keyID][]*fakeKey{}
+	}
+	f.keys[gid][newKey.id] = append(f.keys[gid][newKey.id], newKey)
+}
+
+func (f *fakeEncryptionService) GetEncryptionKey(ctx context.Context, req *enpb.GetEncryptionKeyRequest) (*enpb.GetEncryptionKeyResponse, error) {
+	f.requests.Add(1)
+	userInfo, err := f.authenticator.AuthenticatedUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	gid := groupID(userInfo.GetGroupID())
+
+	// Figure out which key is being requested.
+	var kid keyID
+	if req.GetMetadata().GetId() != "" {
+		kid = keyID(req.GetMetadata().GetId())
+	} else {
+		// If no Key ID was specified, select the key with the highest seq
+		seq := -1
+		for _, keys := range f.keys[gid] {
+			for _, key := range keys {
+				if key.seq > seq {
+					kid = key.id
+					seq = key.seq
+				}
+			}
+		}
+	}
+
+	if kid == "" {
+		return nil, status.NotFoundError("No encryption key available")
+	}
+
+	keys := f.keys[gid][kid]
+	if req.GetMetadata() != nil && req.GetMetadata().GetVersion() >= 0 {
+		for _, key := range keys {
+			if key.version == req.GetMetadata().GetVersion() {
+				return &enpb.GetEncryptionKeyResponse{
+					Key: &enpb.EncryptionKey{
+						Metadata: &enpb.EncryptionKeyMetadata{
+							Id:      string(kid),
+							Version: key.version,
+						},
+						Key: key.key,
+					},
+				}, nil
+			}
+		}
+		return nil, status.NotFoundError("Encryption key version not found")
+	}
+
+	var keyToReturn *fakeKey
+	for _, key := range keys {
+		if keyToReturn == nil {
+			keyToReturn = key
+		} else if key.version > keyToReturn.version {
+			keyToReturn = key
+		}
+	}
+
+	return &enpb.GetEncryptionKeyResponse{
+		Key: &enpb.EncryptionKey{
+			Metadata: &enpb.EncryptionKeyMetadata{
+				Id:      string(kid),
+				Version: keyToReturn.version,
+			},
+			Key: keyToReturn.key,
+		},
+	}, nil
+}
+
+func setup(t *testing.T) (*testauth.TestAuthenticator, interfaces.Crypter, clockwork.FakeClock, *fakeEncryptionService) {
+	te := testenv.GetTestEnv(t)
+	authenticator := testauth.NewTestAuthenticator(testauth.TestUsers(user1, group1, user2, group2, user3, group3))
+	te.SetAuthenticator(authenticator)
+	encryptionService := newFakeEncryptionService(authenticator)
+	grpcServer, runServer, lis := testenv.RegisterLocalGRPCServer(t, te)
+	enpb.RegisterEncryptionServiceServer(grpcServer, encryptionService)
```

Comment:
typically in tests I've seen a new service is constructed via New which allows passing in a mock clock directly, rather than having to put this in the env and hope the service uses the clock from the env